doc/source/user/basics.subclassing.rst

```diff
@@ -31,6 +31,37 @@ things like array slicing.  The complications of subclassing ndarray are
 due to the mechanisms numpy has to support these latter two routes of
 instance creation.
 
+When to use subclassing
+=======================
+
+Besides the additional complexities of subclassing a NumPy array, subclasses
+can run into unexpected behaviour because some functions may convert the
+subclass to a baseclass and "forget" any additional information
+associated with the subclass.
+This means that subclassing can be a convenient approach, but should often be
+avoided in favor of using other interoperability approaches described in
+:ref:`Interoperability with NumPy <basics.interoperability>`.
+
+Subclassing can be a good approach if:
+
+* Your use-case is limited:  If your use-case is limited you may not worry
+  about the surpises if a function ignores the subclass.
+* You do not expect it to be a big surprise if the subclass information is
+  lost silently.  An example is ``np.memmap`` where "forgetting" about
+  data being memory mapped cannot lead to a wrong result.
+  An example of a problematic subclass are NumPy's masked arrays.
+  Converting them to a base class array loses the mask information leading
+  to unexpected results.  With the new interoperability features, we would
+  try to avoid subclassing when implementing masked arrays.
+
+Note that also subclasses may need to study
+:ref:`Interoperability with NumPy <basics.interoperability>`
+if they wish to support more complex use-cases.
+
+``astropy.units.Quantity`` and ``xarray`` are examples for array-like
+objects that interoperate well with NumPy.
```

Comment:
Quantity uses subclasses, though, so it is perhaps not the best example. Maybe pint or pandas would be better choices?

Reply:
True, although `astropy` gets away with it mostly ;).  I am worried that pandas is way too complicated (if even a complete example at all) to direct people to.  But happy to point to `pint` rather than astropy. – the interop page lists a lot of projects as well.

Reply:
I think `Quantity` is good to mention as it is very widely used (I would think more than `pint`, though have no real numbers). It also shows that it is possible to use a subclass but that it is nearly the same amount of work as writing a new class. In fairness, the text should probably state that before `__array_ufunc__` and `__array_function__`, there really was not much of a choice, other than reimplementing the whole numpy API.

p.s. Despite this, I fear astropy has more `ndarray` subclasses; long-standing are `Column` and `MaskedColumn` for table columns, newer are `Distribution` and `MaskedNDArray` (the latter because `MaskedArray` really has its order wrong, putting the mask above the subclass rather than with the data; it was just not possible to make a `MaskedQuantity` with it; see https://docs.astropy.org/en/latest/utils/masked/index.html). It remains true that quite a few more things "just work" as an `ndarray` subclass than as an array mimic.

Reply:
> the text should probably state that before __array_ufunc__ and __array_function__, there really was not much of a choice, other than reimplementing the whole numpy API.

Yeah, I tried to allure to it that when I bring up masked arrays.  I do think that masked array subclassing was the only viable approach at the time.  Let me aim for expanding the main text to be clear about that as well, something like: "While subclassing was the preferred approach to extending NumPy for a very long time, many usecases may not be better served by exclusively using our <Interop feature site>."

(will push/work on this later)